[NewPing] Use Direct GPIO access to reduce jitter on ultrasonic sensors

TD-er opened the pull request:
See: https://github.com/letscontrolit/ESPEasy/issues/4165

- Upgrading the NewPing library to one being 4 years newer ;)
- Re-apply the ESPEasy specific changes to know reason for failing a ping
- Re-apply the more accurate conversion to float cm (or inch)
- Use GPIO Direct access, the same as for Dallas sensors (See #4164 )

Also added fix for timing critical issues for the P005 DHT sensors.
The P005 DHT plugin should now (finally) no longer be dependent on those awfully critical timings as reported by some users.
Those DHT sensors may have some drift in timings based on temperature (for a temperature sensor....)
Now the timings are all relative, so any drift among sensors should no longer matter.

Still to do:
- [ ] Check if several ping tasks can send out pings while other is still active
- [x] Actually test on ESP32 (and ESP8266) to get some idea of the improvements using direct GPIO access
- [ ] Document plugin
- [x] Get official permissing to fork the NewPing library because of changed license. Email has been sent.
- [x] Test Ultrasonic sensor(s)
- [x] Test DHTxx sensors
- [ ] Add AM2320 sensor to the DHTxx plugin (#2942)
- [x] Check effect of reported bug in ESP32 micros(): https://github.com/pstolarz/OneWireNg/issues/38

The new license of NewPing states:
```
// NewPing Library - v1.9.4 - 01/26/2022
//
// AUTHOR/LICENSE:
// Created by Tim Eckel - eckel.tim@gmail.com
// Copyright 2022 License: Forks and derivitive works are NOT permitted without
// permission. Permission is only granted to use as-is for private and
// commercial use. If you wish to contribute, make changes, or enhancements,
// please create a pull request. I get a lot of support issues from this
// library, most of which comes from old versions, buggy forks or sites without
// proper documentation, just trying to wrangle this project together.
```

This is a reasonable requests from the author, but merging this PR will be halted until official permission has been given or suggested changes to the library may have been merged via some pull request to the original library.

Edit:
Permission has been granted via email (within 2 hours :) )

Fixes: #4026
Fixes: #3494

TD-er committed:
[NewPing] Rename library from NewPingESP8266 to NewPing

TD-er committed:
[NewPing] Update Library to 1.9.4

TD-er committed:
[NewPing] Re-apply ESPEasy specific additions to NewPing library

TD-er committed:
[NewPing] Use Direct GPIO access to reduce jitter on ultrasonic sensors

TD-er committed:
[DHT] Use Direct GPIO access for DHTxx sensors

tonhuisman commented on src/_P005_DHT.ino:
```diff
@@ -4,27 +4,15 @@
 
+#include "src/PluginStructs/P005_data_struct.h"
 
 #define PLUGIN_005
```
Think it's missing the `#ifdef USES_P005` here

TD-er replied:
Yep it was indeed missing.

TD-er committed:
[DHT] Remove timing critical aspects from reading DHT sensors

Now the bits are read and an average low duration is computed.
The "high" state duration should differ significant from the "low" state.
Thus if the "high" state duration was longer than the average "low" state, it was a logic "1".

TD-er commented:
Would be really nice if someone could test these plugins.
Neither has been tested by me on real hardware.

The P005 DHT plugin should now (finally) no longer be dependent on those awfully critical timings as reported by some users.
Those DHT sensors may have some drift in timings based on temperature (for a temperature sensor....)
Now the timings are all relative, so any drift among sensors should no longer matter.

TD-er committed:
[DHT] Remove short calls to delayMicroseconds

Based on the information gathered in this issue: https://github.com/pstolarz/OneWireNg/issues/38

toast1234 commented:
I tested on a board with one DHT22 and one DHT11:

Upgrade to - 
ESP_Easy_mega_20220730_normal_ESP8266_4M1M.bin - No values for DHT22 or DHT11. DS18B20 is working

Rollback to -
ESP_Easy_mega_20220616_normal_ESP8266_4M1M.bin - Working with all 3 sensors

Did the upgrade/Rollback twice with same result.

![3](https://user-images.githubusercontent.com/52626255/182013496-43a915f4-d647-479a-bc97-2333deb02242.png)
![4](https://user-images.githubusercontent.com/52626255/182013500-f49b6d04-3db4-4bc5-a404-38e33a06872c.png)

TD-er committed:
[DHT] Add delay right after waking sensor

TD-er committed:
[DHT] Add missing initialization

TD-er committed:
[DHT] Fix using correct index when recording timings + comments update

TD-er committed:
[DHT] Fix GPIO not yet set to input when DHT wake up.

TD-er committed:
[DHT] Fix switching pinMode(INPUT_PULLUP) way too slow

TD-er committed:
[Direct GPIO] Reduce iRAM usage

TD-er committed:
[DHT] Set the GPIO pull-up resistor explicitly

Apparently if you never explicitly set the registers for pull-up or pull-down resistors, these registers may contain random values at boot. No matter if it is a warm or cold boot.

ErikApption commented:
> Would be really nice if someone could test these plugins. Neither has been tested by me on real hardware.

I have issues with couple sensors on my ESP32, I'm willing to give this a try however, I was not able to find a nightly build or simple instructions to build it. If you can point me to one of these, I'd like to test my SR04 and DHT11. Both are acting with recent espeasy and tasmota builds.

TD-er commented:
Both should be in the "normal" build
So if you pick one of [this GH Actions build](https://github.com/letscontrolit/ESPEasy/actions/runs/2809453664) you should have all the latest code in it.

TD-er commented:
By the way, which other sensors are having issues on ESP32 lately?

TD-er commented:
I did test the ultrasonic sensor myself both on ESP32 as well as ESP8266.
Both seem to work fine.

And for fun, here a screenshot of the signals on the trigger and echo pins:
![image](https://user-images.githubusercontent.com/3751318/183309816-f2330a1b-d06c-4803-b6cb-f55b241ca9c0.png)

The 2 signals on the top are digitally sampled.
The bottom 2 signals are the same pins, but measured analog.

You can clearly see the 8 pulses of 25 usec, sent by the sensor (on the analog signal) and the received pulse width on the echo pin.  (25 usec pulse => 40 kHz)

ErikApption commented:
I can confirm same here. Also this change made a huge difference on my DHT11 - it was receiving alot of invalid data before now it seems to be perfect again.

TD-er commented:
Have you also tried the Dallas sensors?
Those also work very very well now. :)

ErikApption commented:
Ok so it is not totally resolved... I tested two different SR04 sensors and same result.

I'm getting quite a bit of these errors. On older ESP8266, ESPEasy + SR04, I was getting proper measurements - aka 2+ meters but with this new version I'm only getting readings under 2m. Also it would be nice if we could get a null or -1 for errors instead of 0.

500536: EVENT: GarageDoor#Distance=0.00
501506: ULTRASONIC : TaskNr: 1 Distance: 0.00 cm Error: Echo too late, maximum distance exceeded
501535: EVENT: GarageDoor#Distance=0.00
502507: ULTRASONIC : TaskNr: 1 Distance: 0.00 cm Error: Echo too late, maximum distance exceeded

TD-er commented:
Hmm will check the differences among library versions as I did not change anything related to the max. distance.
Over 38 msec it should -as far as I know- consider it a "no ping received" and below there is a limit set by the user which is then given to the library as an upper limit.

TD-er commented:
I did a quick test here and it measured upto 3m55 when pointing it to the ceiling, which is about right.  (it is a triangle shape so it is hard to point it exactly into the top corner)
Make sure you have the distance upper limit (Max Distance) set to allow over 200 cm.
![image](https://user-images.githubusercontent.com/3751318/183582137-ed1d3f0f-a80d-493c-b245-653fa2ece227.png)
